### ISSUE TITLE
Calypso theme page shows incorrect theme for Storefront child theme "Bistro"

### ISSUE THREAD

**Issue by mikeicode** [Automattic/wp-calypso#54411]
When an AT site has the [Storefront child theme Bistro](https://woocommerce.com/products/bistro/) activated the incorrect theme is shown in Calypso. This has resulted in confusion in support.

### Steps to reproduce the behavior

1. Activate Bistro on an AT site: https://woocommerce.com/products/bistro/
2. View the active theme in Calypso: https://wordpress.com/themes/YOUR-SITE.COM
2. View the active theme in WP Admin: https://YOUR-SITE.COM/wp-admin/themes.php

#### What I expected to happen

WP Admin shows correct theme:

<img width="991" alt="Screen Shot 2021-07-09 at 08 50 54" src="https://user-images.githubusercontent.com/10353785/125098535-17e45e00-e0a5-11eb-89ff-fb4cd655738c.png">

Calypso should show the same correct theme.

#### What actually happened

Calypso shows incorrect theme:

<img width="1233" alt="Screen Shot 2021-07-09 at 08 48 55" src="https://user-images.githubusercontent.com/10353785/125098602-29c60100-e0a5-11eb-9323-58533ea0343a.png">

### Context

This resulted in confusion during multiple interactions, the latest being:

30544088-hc

#### Browser / OS version

Mac 10.15.5 - Firefox (probably not related to browser or OS)

#### Is this specific to the applied theme? Which one?

https://woocommerce.com/products/bistro/

#### Does this happen on simple or atomic sites or both?

AT

#### Level of impact (Does it block purchases? Does it affect more than just one site?)

This theme isn't included with the eCommerce plan, so it only affects anyone that purchases the theme.

**Comment by kosiew** [comment]
Hi @mikeicode ,

Thanks for reporting this and the detailed steps.

Managed to reproduce this on my test site too.
Noted that this is specific to Calypso, not in wp-admin.

**Comment by mreishus** [comment]
I'm able to reproduce this. After purchasing and installing bistro on my atomic site, the theme showcase "Current Theme" card, and the info page for that theme, display information about the wrong Bistro theme. It shows a picture of a pizza (wrong theme), not a picture of vegetables (correct theme). The description mentions being a child theme of Sydney (wrong theme), not being a storefront theme (correct theme).


The core of the conflict seems to be a conflict with an old theme called Bistro. [Here it is on wp-themes.com](https://wp-themes.com/bistro/). Here's [an API request](https://api.wordpress.org/themes/info/1.1/?action=theme_information&request%5Bfields%5D%5Bextended_author%5D=true&request%5Bslug%5D=bistro) that pulls bistro's information.
On the data layer of themes in calypso, we identify themes with a "Theme ID". In this case, it's the string `bistro`. I believe the code assumes this is a unique identifier when it is not.
It always queries WPCOM for information on the theme. If it finds information, it stops. If it can't find anything on WPCOM, it asks WPORG. If it finds information, it stops. If it can't find info on WPCOM or WPORG, only then does it ask the sites itself (which for a jetpack or atomic site, would retrieve the information on disk). [Code block.](https://github.com/Automattic/wp-calypso/blob/c53332a1951a72e332b8eb0c6bbd9ca36228d38d/client/components/data/query-canonical-theme/index.jsx#L17-L19)
In this case, it is finding the theme information on either WPCOM or WPORG, so it does not bother to ask the atomic site itself to return the theme details.
If I manually disable the WPCOM and WPORG theme checks, then it works:
![2021-08-03_16-11](https://user-images.githubusercontent.com/937354/128087917-77639d0b-64e3-4b14-b02c-f852b845ad92.png)
![2021-08-03_16-11_1](https://user-images.githubusercontent.com/937354/128087930-d4566ba8-4745-4438-b742-596e12c08648.png)

This does seem tricky. I will need to spend some time to think of a clean solution.

**Comment by mreishus** [comment]
The first thing I tried was the straight forward approach of 'always preferring theme information from a jetpack/atomic site if it exists':
```diff
diff --git a/client/components/data/query-canonical-theme/index.jsx b/client/components/data/query-canonical-theme/index.jsx
index 14da58d6a5..7d4d2476f3 100644
--- a/client/components/data/query-canonical-theme/index.jsx
+++ b/client/components/data/query-canonical-theme/index.jsx
@@ -16,7 +16,7 @@ const QueryCanonicalTheme = ( { siteId, themeId, isWpcom, isWporg } ) => (
        <Fragment>
                <QueryTheme themeId={ themeId } siteId="wpcom" />
                { ! isWpcom && <QueryTheme themeId={ themeId } siteId="wporg" /> }
-               { ! isWpcom && ! isWporg && siteId && <QueryTheme themeId={ themeId } siteId={ siteId } /> }
+               { siteId && <QueryTheme themeId={ themeId } siteId={ siteId } /> }
        </Fragment>
 );
 
diff --git a/client/state/themes/selectors/get-canonical-theme.js b/client/state/themes/selectors/get-canonical-theme.js
index 648d61f790..a6d7cd790a 100644
--- a/client/state/themes/selectors/get-canonical-theme.js
+++ b/client/state/themes/selectors/get-canonical-theme.js
@@ -22,6 +22,6 @@ import 'calypso/state/themes/init';
  * @returns {?object}         Theme object
  */
 export function getCanonicalTheme( state, siteId, themeId ) {
-       const source = find( [ 'wpcom', 'wporg', siteId ], ( s ) => getTheme( state, s, themeId ) );
+       const source = find( [ siteId, 'wpcom', 'wporg' ], ( s ) => getTheme( state, s, themeId ) );
        return getTheme( state, source, themeId );
 }
```

Unfortunately, that leads to a degraded experience when viewing theme information for default, bundled themes like twentytwentyone on jetpack/atomic sites:
![2021-08-04_10-32](https://user-images.githubusercontent.com/937354/128213434-c76da5de-2682-4373-92af-5db8ac014707.png)
The information we keep in the wpcom gallery is much richer and includes all fields.

Therefore, I have added a list of "conflicting themeids" which are the only ones that prefer to find their information on jetpack/atomic: https://github.com/Automattic/wp-calypso/pull/55162
This seems overly manual and somewhat hackish, however, I an unable to find another way that fixes this problem without negatively affecting the theme info experience on other themes. Open to any other suggestions.

**Comment by mreishus** [comment]
We've added a bistro specific workaround to calypso:
![2021-08-10_16-10](https://user-images.githubusercontent.com/937354/128935569-6aa216f5-e07c-4c7a-a7c9-467aaa1527ee.png)

For bistro, it will prefer looking up theme information on local files, if they exist, in priority over the theme gallery. That's done here: ( https://github.com/Automattic/wp-calypso/pull/55162 ).

It's not the most elegant of solutions. I raised the issue of the Bistro slug conflict here: 10982-gh-Automattic/woocommerce.com . If we're able to rename the bistro theme (and I don't know the feasibility of that), we could remove the workaround added above.
